### ISSUE TITLE
Gradle 8 RC2 runs out of metaspace

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#23698]
### Expected Behavior
Works as before.

### Current Behavior
I'm consistently getting
```
    The Daemon will expire after the build after running out of JVM Metaspace.
    The project memory settings are likely not configured or are configured to an insufficient value.
    The daemon will restart for the next build, which may increase subsequent build times.
    These settings can be adjusted by setting 'org.gradle.jvmargs' in 'gradle.properties'.
    The currently configured max heap space is '512 MiB' and the configured max metaspace is '256 MiB'.
    For more information on how to set these values, visit the user guide at https://docs.gradle.org/8.0-rc-2/userguide/build_environment.html#configuring_jvm_memory
    To disable this warning, set 'org.gradle.daemon.performance.disable-logging=true'.
    Daemon will be stopped at the end of the build after running out of JVM Metaspace
```

### Context
I have a project which is a set of Gradle plugins. This project is using GradleRunner from Gradle TestKit to execute builds in temporary folders. There's complex CI setup to run on all kinds of Android Gradle Plugin and Gradle combinations (Kotlin and Java versions are fixed for each combo). Here's an [example workflow](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4029307232) from `master`.

I'm in the process of adopting Gradle 8 RC 1/2 in [PR](https://github.com/TWiStErRob/net.twisterrob.gradle/pull/357). In this PR I'm trying get the latest AGP 7.4.0 + Gradle 8 RC2 combination to be green. Locally it's OK, but on CI it keeps timing out, because of GradleRunner daemons running out of metaspace.

This means that I'm not directly observing this behavior, but it shows up on trivial near-empty projects in integration tests. I'm in the process of trying narrow it down to prove that Gradle 8 RC 1/2 is the root cause or another change in the PR (it's a lot of commits 😞). In the meantime I'm reporting this in case there are others experiencing the same, or you have any idea why it could be.

### Steps to Reproduce 
Sadly the only thing I have is a branch which consistently fails on GitHub Actions. I was able to reproduce it once locally, it's not as consistent.

### Your Environment
Gradle 8.0 [RC2](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4032142236/jobs/6931841029) is definitely failing, so does [RC1](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4032309412/jobs/6932107642). Gradle 7.5.1 and Gradle 7.6 is not affected at all as can be seen [here in the sidebar](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4032142236/jobs/6931841029).

Build scan URL: Not possible, Gradle build doesn't finish, GitHub Actions terminates after timeout.

**Comment by TWiStErRob** [comment]
I was able to "fix" all the tests that were failing by increasing metaspace from 256M to 384M in these tests:
 * AndroidSigningPluginIntgTest.kt
 * AndroidMinificationPluginIntgTest.kt
 * KotlinPluginIntgTest.kt
 * JavaPluginIntgTest.kt

The commonalities between them is that most of them do Kotlin or Java compilation.

---

In the meantime I also noticed that I've [given](https://github.com/TWiStErRob/net.twisterrob.gradle/pull/357/commits/031f5562700cceb773924a1f12a1e66de61222a2) [more](https://github.com/TWiStErRob/net.twisterrob.gradle/pull/357/commits/f152c70ae2378d304d65cd933da302b00cac64e7) memory to [small manual integration example projects](https://github.com/TWiStErRob/net.twisterrob.gradle/tree/a23facc6026a347b877ff018c433fbc22e7cea26/docs/examples) too, but strangely I haven't been able to reproduce it on those projects when I was investigating this issue. These projects apply AGP and one of my plugins, otherwise empty.

**Comment by ljacomet** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed, response time may vary.

**Comment by TWiStErRob** [comment]
@ljacomet I don't think it's only plugin development affected, as all the builds running are simple projects. I have an idea to confirm this.

**Comment by TWiStErRob** [comment]
Potentially related https://youtrack.jetbrains.com/issue/KT-56093

**Comment by MGaetan89** [comment]
Might be fixed by https://github.com/gradle/gradle/pull/23749?

**Comment by TWiStErRob** [comment]
RC3 milestone is all closed, I think they'll prepare it soon.

**Comment by TWiStErRob** [comment]
And it's out now https://docs.gradle.org/8.0-rc-3/release-notes.html

**Comment by SimonMarquis** [comment]
The warning message still appears on rc-3 on a trivial build https://github.com/SimonMarquis/SealedObjectInstances/actions/runs/4103117251/jobs/7076901164#step:4:125

**Comment by TWiStErRob** [comment]
I confirmed it too, #23749 does not fix this.

https://github.com/TWiStErRob/net.twisterrob.gradle/pull/399
 * RC2: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/399/commits/7c8052ee9da068e021840647e250c838d3b0e74d -> [timeout because tests thrash after `running out of JVM Metaspace`](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4103077405/jobs/7076818554)
 * RC3: https://github.com/TWiStErRob/net.twisterrob.gradle/pull/399/commits/385830fbd5312035db6fdab448ad4252b90cbc53 -> [timeout because tests thrash after `running out of JVM Metaspace`](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4103660385/jobs/7078581119)

**Comment by TWiStErRob** [comment]
@ljacomet would changing the default be a long shot?
```diff
-org.gradle.jvmargs=-Xmx512M -XX:MaxMetaspaceSize=256M
+org.gradle.jvmargs=-Xmx512M -XX:MaxMetaspaceSize=384M
```
(Related docs: https://github.com/gradle/gradle/blame/v8.0.0-RC3/subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc#L104)

**Comment by SimonMarquis** [comment]
My project includes tests which compiles Kotlin source code snippets (in task `:processor:test`) (using [tschuchortdev/kotlin-compile-testing](https://github.com/tschuchortdev/kotlin-compile-testing)), so this might be the reason why we see these logs here.

**Comment by TWiStErRob** [comment]
> so it makes me think that either the heap size isn't what we think it is or we're misreporting the memory situation

I don't think there's a misreport, see my log below. There are just more than 256M classes on the classpath, I guess; or some classloader leaks that shouldn't.

> I tried reproducing this locally and I don't see that warning. 

I can't repro it locally either, but on GitHub Actions I have a consistent failure. I'm trying to minimize a repro. Also note that [@SimonMarquis' build](https://github.com/gradle/gradle/issues/23698#issuecomment-1419025755) is also running out of metaspace on the CI. @SimonMarquis can you repro locally consistently?

> The build scan seems to say the max heap memory is 512MB, which is our default, so misreporting seems more likely.

The heap is 512, that's fine, the problem is the metaspace, which is 256... I have diagnostic info in my tests to confirm that everything is as expected:
```java
// from the test running the testkit build:
Test Java: Oracle Corporation OpenJDK Runtime Environment 11.0.2 (11.0.2+9 2019-01-15) from P:\tools\lang\java-11.0.2-x64-openjdk.
Requesting Gradle 8.0-rc-5 in worker #1 at C:\Users\TWiStEr\AppData\Local\Temp\.gradle-test-kit-TWiStEr-1.

// beginning of build

// from an init script's rootProject { }
Gradle 8.0-rc-5 worker #1 at C:\Users\TWiStEr\AppData\Local\Temp\.gradle-test-kit-TWiStEr-1 with P:\caches\gradle\wrapper\dists\gradle-8.0-rc-5-all\7yhhfnupylrv44qmug3cozi6u\gradle-8.0-rc-5.
Gradle Java: Oracle Corporation OpenJDK Runtime Environment 11.0.2 (11.0.2+9 2019-01-15) from P:\tools\lang\java-11.0.2-x64-openjdk.

// from an init script's rootProject { }
Gradle memory: heap = 79MB/512MB (432MB free), metaspace = 122MB/256MB (133MB free).

Running `gradle assembleRelease`
// build output

// from an init script's buildFinished { }
Gradle memory: heap = 143MB/512MB (368MB free), metaspace = 241MB/256MB (14MB free).
```

This is a local (Windows) execution, you can see it barely passes. On CI (linux) something must be different.

**Comment by TWiStErRob** [comment]
I managed to reduce it to a smaller example:
CI: https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4150809580/jobs/7180706136
Steps: https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/4150809580/workflow#L5-L27
Files: https://github.com/TWiStErRob/net.twisterrob.gradle/tree/gradle_meta/docs/examples/release

The repro has no dependencies other than `mavenCentral()`.

It reproduces consistently on CI, but not locally. I think this might be because of the fashion of CI, that it starts from scratch every time, and for example it needs to generate gradle-api/dsl jars.

Locally I see this warning / run into metaspace quite often now that more of my projects are using Gradle 8 RC. (I switch between them multiple times a day.)

@big-guy what could we procure to move this along and figure out what's wrong? When I get a metaspace OOM, it dumps heap, but the `.hprof` file doesn't contain Metaspace 🙄

**Comment by TWiStErRob** [comment]
Ok will do, what has changed?

**Comment by eskatos** [comment]
Looking at it, nothing changed wrt memory since RC5.
We'll get our folks to look at your reproducer.

**Comment by lwasyl** [comment]
There's another open issue that hints at KAPT or code involved when KAPT runs: https://youtrack.jetbrains.com/issue/KT-55831/

**Comment by TWiStErRob** [comment]
@lwasyl I think it's simply the fact that AGP and KGP have too many classes that are loaded. I just don't understand why Gradle 7.6 to 8.0 would trigger this. Maybe has to do with a default Kotlin 1.8 bundle? @eskatos do you have some tooling to git bisect the reproducer in some way? Or we're just going to [accept](https://github.com/gradle/gradle/pull/24080) the growth as "business as usual"?

**Comment by adammurdoch** [comment]
@TWiStErRob We're still looking into what we can do about this. It does looks like there are many, many more KGP classes being loaded in Gradle 8.0 when Kotlin DSL scripts are compiled and this can push the metaspace usages over the limit. As to what we can do about this, let's see.

Thanks for the reproducer, btw, it has been extremely useful.

**Comment by ljacomet** [comment]
This will be "resolved" in Gradle 8.0.2 by raising  default metaspace, similar to [this suggestion](https://github.com/gradle/gradle/issues/23698#issuecomment-1419105462).

Note that the Gradle team has plans to look at memory consumption in upcoming 8.x releases.

**Comment by ZacSweers** [comment]
As a general rule, anything involving PSI/UAST needs a lot more metaspace. Especially tasks like android lint, the gradle default has been prohibitively low and you should raise it significantly in your projects. We set it to **8gb** on CI for our lint jobs at Slack.

**Comment by TWiStErRob** [comment]
Yeah, for normal multi-module actual projects I set it higher as well, but trivial/empty projects should run on default, shouldn't they? Also, my discovery came when I was running integration tests in unit tests, increasing those to 8G, and running a 4-8 of them in parallel would be fun 😅.

**Comment by ZacSweers** [comment]
Gradle's defaults are not set specifically for android projects. I think it's important to differentiate your expectations, as gradle is building for a much wider spectrum of projects.

**Comment by TWiStErRob** [comment]
I agree with that, but there's also an expectation that things that were building before, will build on future platforms too; Gradle is extremely good at this. I had the same test running on Gradle 4-8, and it only failed on latest.

If the platform takes up 50-100MB of extra metaspace (probably acceptable considering the 4-version Kotlin jump), then the default is increased to accommodate that. I think this is why they bumped it up in the end. I'm sure there exist other combinations of plugins where this might come up (e.g. Spring, GraphQL, Kotlin, Dokka, Detekt, ktfmt)

Actually, I'm surprised that this didn't come up in internal integration tests, as far as I know Gradle runs AGP-specific tests.

**Comment by mklueh** [comment]
> This will be "resolved" in Gradle 8.0.2 by raising default metaspace, similar to [this suggestion](https://github.com/gradle/gradle/issues/23698#issuecomment-1419105462).
> 
> Note that the Gradle team has plans to look at memory consumption in upcoming 8.x releases.

Hi. I'm running into this issue with 8.0.2 - are you sure this was resolved?

**Comment by TWiStErRob** [comment]
@mklueh it'll never really be fully resolved, as Zac said as well. There'll always be project which need more heap and meta space. Gradle added 128MB extra metaspace **to the default** as a workaround to reduce the chance of this happening on trivial projects. Also as you quoted, the story doesn't end here, there'll be more investigations.

Is it possible you're overriding the jvmargs, because this change won't affect anyone with explicit memory, but the problem exists because the memory usage grew overall, so you might need to increase it a bit as well. (Just guessing here, as I don't know on what project you experienced this.)
